File: apps/tabtin-ios/Tabtin/Core/IM/LastConversationMessageReplayCache.swift
import Foundation

struct LastConversationMessageReplayCache<Message> {
    private var messagesByConversationId: [String: Message] = [:]

    mutating func remember(_ message: Message, for conversationId: String) {
        messagesByConversationId[conversationId] = message
    }

    func replay(for conversationId: String) -> Message? {
        messagesByConversationId[conversationId]
    }

    mutating func clear(conversationId: String) {
        messagesByConversationId.removeValue(forKey: conversationId)
    }

    mutating func clearAll() {
        messagesByConversationId.removeAll()
    }
}

File: apps/tabtin-android/app/src/main/java/com/tabtin/mobile/data/im/LastConversationMessageReplayCache.kt
package com.tabtin.mobile.data.im

/**
 * 会话目录的 `lastMessage` 常比详情 listener / history 更早可见。
 * 缓存最新一条，供详情订阅时无副作用地 replay，避免入口摘要已更新但聊天停在旧消息。
 */
internal class LastConversationMessageReplayCache<Message> {
    private val messagesByConversationId = mutableMapOf<String, Message>()

    fun remember(message: Message, conversationId: String) {
        if (conversationId.isBlank()) return
        messagesByConversationId[conversationId] = message
    }

    fun replay(conversationId: String): Message? =
        messagesByConversationId[conversationId]

    fun clear(conversationId: String) {
        messagesByConversationId.remove(conversationId)
    }

    fun clearAll() {
        messagesByConversationId.clear()
    }
}
